R-package/R/xgb.Booster.R

```diff
@@ -129,7 +129,7 @@ xgb.Booster.complete <- function(object, saveraw = TRUE) {
     stop("argument type must be xgb.Booster")
 
   if (is.null.handle(object$handle)) {
-    object$handle <- xgb.Booster.handle(modelfile = object$raw)
```

Comment:
I'm confused as to how this line is not currently working as desired. The `object$handle` attribute is assigned the return value of `xgb.Booster.handle`, right? Why do we need to also pass `object$handle` as an argument to `xgb.Booster.handle` also?

It seems similar to writing
```cpp
double foobar(double x, const double* y) {
  const double new_val = x + 1;
  *y = new_val;
  return new_val;
}
...
double v;
v = foobar(1.0, &v);
```

Reply:
It's not that it isn't working, but that it won't update the R external pointer object in-place, creating a new one instead. If passed this way (as in the PR), it will just update an existing R object (which holds a null pointer).

E.g.
```r
new_obj <- readRDS(...)
predict(new_obj, new_data) # <- got de-serialized
predict(new_obj, more_data) # <- again de-serialized
...
```
vs.
```r
new_obj <- readRDS(...)
predict(new_obj, new_data) # <- got de-serialized
predict(new_obj, more_data) # <- already de-serialized
```

Currently, when a C++ object gets de-serialized after passing an object with a null pointer handle, the R pointer object to it will be short-lived as there will not be any R object tied to it (it will not be reachable after the function call ends, but it is not immediately garbage-collected).

Since the R pointer object (which is now no longer assigned to anything) is itself small, it's a less likely candidate to get garbage-collected by R, thus postponing C++ object destruction and increasing memory usage when one inadvertenly uses a model with a null handle.

Reply:
I'm still confused as to why calling `predict()` twice will trigger deserialization twice. If `object$handle` is assigned a reference to the new handle, the `object` should still have the reference to the new handle when `predict()` is called the second time?

Reply:
`object$handle` is an element of an R list, which are copy-on-write, not pass-by-reference. Thus, it will not get updated outside of the function call.

Reply:
> R list, which are copy-on-write, not pass-by-reference

Wow I had no idea. I was thinking R lists are like a Python object with attributes, but I was mistaken.